processor/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,504 @@
+
+	p.store = store.NewStore(p.config.Store.TTL, p.config.Store.MaxItems, p.onComplete, p.onExpire)
+
+	exporters := host.GetExporters()
+
+	// The available list of exporters come from any configured metrics pipelines' exporters.
+	for k, exp := range exporters[config.MetricsDataType] {
+		metricsExp, ok := exp.(component.MetricsExporter)
+		if k.String() == p.config.MetricsExporter && ok {
+			p.metricsExporter = metricsExp
+			break
+		}
+	}
+
+	if p.metricsExporter == nil {
+		return fmt.Errorf("failed to find metrics exporter: %s",
+			p.config.MetricsExporter)
+	}
+
+	// TODO: Consider making this configurable.
+	go p.cacheLoop(time.Minute)
+
+	p.logger.Info("Started servicegraphsprocessor")
+	return nil
+}
+
+func (p *processor) Shutdown(_ context.Context) error {
+	p.logger.Info("Shutting down servicegraphsprocessor")
+	close(p.shutdownCh)
+	return nil
+}
+
+func (p *processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (p *processor) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	if err := p.aggregateMetrics(ctx, td); err != nil {
+		return fmt.Errorf("failed to aggregate md: %w", err)
+	}
+
+	md, err := p.buildMetrics()
+	if err != nil {
+		return fmt.Errorf("failed to build metrics: %w", err)
+	}
+
+	// Skip empty metrics.
+	if md.MetricCount() == 0 {
+		return nil
+	}
+
+	// Firstly, export md to avoid being impacted by downstream trace processor errors/latency.
+	if err := p.metricsExporter.ConsumeMetrics(ctx, md); err != nil {
+		return err
+	}
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *processor) aggregateMetrics(ctx context.Context, td ptrace.Traces) (err error) {
+	var (
+		isNew             bool
+		totalDroppedSpans int
+	)
+
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rSpans := rss.At(i)
+
+		rAttributes := rSpans.Resource().Attributes()
+
+		serviceName, ok := findServiceName(rAttributes)
+		if !ok {
+			// If service.name doesn't exist, skip processing this trace
+			continue
+		}
+
+		scopeSpans := rSpans.ScopeSpans()
+		for j := 0; j < scopeSpans.Len(); j++ {
+			spans := scopeSpans.At(j).Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				connectionType := store.Unknown
+
+				switch span.Kind() {
+				case ptrace.SpanKindProducer:
+					// override connection type and continue processing as span kind client
+					connectionType = store.MessagingSystem
+					fallthrough
+				case ptrace.SpanKindClient:
+					traceID := span.TraceID().HexString()
+					key := buildEdgeKey(traceID, span.SpanID().HexString())
+					isNew, err = p.store.UpsertEdge(key, func(e *store.Edge) {
+						e.TraceID = traceID
+						e.ConnectionType = connectionType
+						e.ClientService = serviceName
+						e.ClientLatencySec = float64(span.EndTimestamp()-span.StartTimestamp()) / float64(time.Millisecond.Nanoseconds())
+						e.Failed = e.Failed || span.Status().Code() == 2
```

Comment:
What's 2 here? Is there a constant somewhere that can be used here?

Reply:
Changed to use variable